zerver/lib/rate_limiter.py

```diff
@@ -19,37 +22,99 @@
 
 client = get_redis_client()
 rules = settings.RATE_LIMITING_RULES  # type: List[Tuple[int, int]]
-def _rules_for_user(user):
-    # type: (UserProfile) -> List[Tuple[int, int]]
-    if user.rate_limits != "":
-        result = []  # type: List[Tuple[int, int]]
-        for limit in user.rate_limits.split(','):
-            (seconds, requests) = limit.split(':', 2)
-            result.append((int(seconds), int(requests)))
-        return result
-    return rules
 
 KEY_PREFIX = u''
 
+class RateLimitedObject(object):
+    def get_keys(self):
+        # type: () -> List[Text]
+        key_fragment = self.key_fragment()
+        return ["{}ratelimit:{}:{}".format(KEY_PREFIX, key_fragment, keytype)
+                for keytype in ['list', 'zset', 'block']]
+
+    def key_fragment(self):
+        # type: () -> Text
+        raise NotImplemented
+
+    def rules(self):
+        # type: () -> List[Tuple[int, int]]
+        raise NotImplemented
+
+    def check_request_is_ratelimited(self, entity, request):
+        # type: (RateLimitedObject, HttpRequest) -> None
+        raise NotImplemented
+
+    @classmethod
+    def process_response(cls, request, response):
+        # type: (HttpRequest, HttpResponse) -> None
+        raise NotImplemented
+
+class RateLimitedUser(RateLimitedObject):
+    def __init__(self, user, domain='all'):
+        # type: (UserProfile, Text) -> None
+        self.user = user
+        self.domain = domain
+
+    def key_fragment(self):
+        # type: () -> Text
+        return "{}:{}:{}".format(type(self.user), self.user.id, self.domain)
+
+    def rules(self):
+        # type: () -> List[Tuple[int, int]]
+        if self.user.rate_limits != "":
+            result = []  # type: List[Tuple[int, int]]
+            for limit in self.user.rate_limits.split(','):
+                (seconds, requests) = limit.split(':', 2)
+                result.append((int(seconds), int(requests)))
+            return result
+        return rules
+
+    def check_request_is_ratelimited(self, entity, request):
```

Comment:
This method is buggy -- it's written as a method on one RateLimitedObject, but then it takes another as a parameter, and alternates referring to one or the other interchangeably to do its work. Basically it assumes that the argument passed as `entity` is the same as `self`, but that's not specified in its interface at all.

I think this doesn't belong as a method here, or in this source file -- it's working specifically on an `HttpRequest`, which tangles up this otherwise pretty well-isolated module with the specifics of handling an HTTP request at the outer layer of Django. I'd rather keep it in `zerver/decorator.py`, which is largely about that stuff already.